src/ServerHandler.cpp

```diff
@@ -1,8 +1,153 @@
+    }
+}
+
+std::string ServerHandler::extractBookName(std::vector <std::string> pieces, size_t startPos, size_t endPos) {
+    std::string book;
+    // the || is unnecessary with right usage, added here only because the sake of style.
+    for (size_t i = startPos; i < endPos || i < pieces.size(); i++) {
+        book += pieces[i];
+    }
+    return book;
+}
+
+std::string ServerHandler::booksString(std::vector <std::string> books) {
+    std::string body = _inventory.getUsername() + ":";
+    for (std::string book : books) {
+        body += book;
+        body += ",";
+    }
+    body.substr(0, body.size() - 1); // instead of adding a check for last book, just remove last ','
+    return body;
+}
+
+void ServerHandler::parseMessageFrame(Frame messageFrame) {
+    std::string genre = messageFrame.getHeaderVal("destination");
+    std::vector <std::string> subStrs;
+    std::string body = messageFrame.getBody();
+    split(body, subStrs, " ");
+    if (subStrs[1] == "wish") {
+        std::string book = extractBookName(subStrs, 3, subStrs.size()); // starting from the word 'borrow'
+        if (_inventory.isInInventory(genre, book)) {
+            Frame ansFrame = Frame(SEND);
+            ansFrame.addHeader("destination", genre);
+            ansFrame.setBody(_inventory.getUsername() + " has " + book);
+            _connectionHandler->sendFrameAscii(ansFrame.toString(), STOMP_DELIMITER.c_str()[0]);
+        }
+    }
+    else if (subStrs[0] == "Returning" && subStrs[subStrs.size() - 1] == _inventory.getUsername()) {
+        // starting after 'Returning', ending before 'to'
+        std::string book = extractBookName(subStrs, 1, subStrs.size() - 2);
+        _inventory.addBook(genre, book);
+    }
+    else if (subStrs[0] == "Taking" && subStrs[subStrs.size() - 1] == _inventory.getUsername()) {
+        // starting after 'Taking', ending before 'from'
+        std::string book = extractBookName(subStrs, 1, subStrs.size() - 2);
+        _inventory.removeFromInventory(genre, book);
+    }
+    else if (body == "book status") {
+        Frame ansFrame = Frame(SEND);
+        ansFrame.addHeader("destination", genre);
+        std::string body = booksString(_inventory.getGenreBooks(genre));
+        ansFrame.setBody(body);
+        _connectionHandler->sendFrameAscii(ansFrame.toString(), STOMP_DELIMITER.c_str()[0]);
+    }
+    else if (subStrs[1] == "has") {
+        std::string book = extractBookName(subStrs, 2, subStrs.size()); //staring after the word 'has'
+        if (_inventory.isInWishList(genre, book)) {
+            std::string lender = subStrs[0];
+            Frame ansFrame = Frame(SEND);
+            ansFrame.addHeader("destination", genre);
+            ansFrame.setBody("Taking " + book + " from " + lender);
+            _connectionHandler->sendFrameAscii(ansFrame.toString(), STOMP_DELIMITER.c_str()[0]);
+            _inventory.borrowBook(book, lender);
+            _inventory.removeFromWishList(genre, book);
+        }
```

Comment:
Is this fine without synchronization? It seems likely to be fine but be careful of what happens if the user does something weird here. Perhaps adding a book that was in the wishList to the inventory while you are getting that book from someone else (If this is a situation that can even happen).

Reply:
I won't even check for it. if a user asks for a book, then add it by itself, he is stupid, and it is for is own risk